core/src/test/scala/unit/kafka/server/KafkaApisTest.scala

```diff
@@ -9510,6 +9620,25 @@ class KafkaApisTest extends Logging {
       requestMetrics, envelope = None)
   }
 
+  private def buildRequestWithPrincipal(request: AbstractRequest,
+                                        principal: KafkaPrincipal,
+                                        clientAddress: InetAddress,
+                                        listenerName: ListenerName = ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT),
+                                        fromPrivilegedListener: Boolean = false,
+                                        requestHeader: Option[RequestHeader],
+                                        requestMetrics: RequestChannelMetrics = requestChannelMetrics): RequestChannel.Request = {
+    val buffer = request.serializeWithHeader(
+      requestHeader.getOrElse(new RequestHeader(request.apiKey, request.version, clientId, 0)))
+
+    // read the header from the buffer first so that the body can be read next from the Request constructor
+    val header = RequestHeader.parse(buffer)
+    val context = new RequestContext(header, "1", clientAddress, Optional.empty(),
+      principal, listenerName, SecurityProtocol.SSL,
+      ClientInformation.EMPTY, fromPrivilegedListener, Optional.of(kafkaPrincipalSerde))
+    new RequestChannel.Request(processor = 1, context = context, startTimeNanos = 0, MemoryPool.NONE, buffer,
+      requestMetrics, envelope = None)
+  }
```

Comment:
Seems existing `buildRequest` method can be overloaded with this method. So we have common code.